cpp/fanos_shared.c

```diff
@@ -0,0 +1,212 @@
+#include "cpp/fanos_shared.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <stdarg.h>  // va_list, etc.
+#include <stdio.h>   // vfprintf
+#include <stdlib.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+#define NUM_FDS 3
+#define SIZEOF_FDS (sizeof(int) * NUM_FDS)
+
+// error string helper
+static void set_err(char** dst, const char* msg) {
+  *dst = (char*)malloc(strlen(msg) + 1);
```

Comment:
Hm will this cause a leak?

How about we just say that all the error strings must be GLOBAL CONSTANTS?  And I think they are.  There is no formatting

Then we can just assign them directly, and avoid malloc?

( It would be  nice to have ASAN test this, but I'm not sure how to do it with Python's distutils build.  Should  be pretty possible though )

Reply:
Ah good question... The user of `fanos_*()` owns `*dst`, so I figured something in python would handle freeing it once the exception was destroyed. The docs for `PyErr_SetString()` (https://docs.python.org/3/c-api/exceptions.html#c.PyErr_SetString) aren't super clear on how the `const char*` arg is treated. Though, it would definitely be odd if it were freeing a `const` that it didn't own... I like the globals idea, though. I'll check that it's safe to do and switch to it